src/uucore/src/lib/features/checksum.rs

```diff
@@ -502,208 +590,243 @@ fn identify_algo_name_and_length(
     Some((algorithm, bits))
 }
 
-/***
- * Do the checksum validation (can be strict or not)
-*/
+/// Parses a checksum line, detect the algorithm to use, read the file and produce
+/// its digest, and compare it to the expected value.
+///
+/// Returns `Ok(bool)` if the comparison happened, bool indicates if the digest
+/// matched the expected.
+/// If the comparison didn't happen, return a `LineChecksumError`.
 #[allow(clippy::too_many_arguments)]
-pub fn perform_checksum_validation<'a, I>(
-    files: I,
-    strict: bool,
-    status: bool,
-    warn: bool,
-    binary: bool,
-    ignore_missing: bool,
-    quiet: bool,
-    algo_name_input: Option<&str>,
-    length_input: Option<usize>,
-) -> UResult<()>
-where
-    I: Iterator<Item = &'a OsStr>,
-{
-    // if cksum has several input files, it will print the result for each file
-    for filename_input in files {
-        let mut correct_format = 0;
-        let mut properly_formatted = false;
-        let mut res = ChecksumResult::default();
-        let input_is_stdin = filename_input == OsStr::new("-");
-
-        let file: Box<dyn Read> = if input_is_stdin {
-            // Use stdin if "-" is specified
-            Box::new(stdin())
-        } else {
-            match get_input_file(filename_input) {
-                Ok(f) => f,
-                Err(e) => {
-                    // Could not read the file, show the error and continue to the next file
-                    show_error!("{e}");
-                    set_exit_code(1);
-                    continue;
-                }
+fn process_checksum_line(
+    filename_input: &OsStr,
+    line: &OsStr,
+    i: usize,
+    chosen_regex: &Regex,
+    is_algo_based_format: bool,
+    res: &mut ChecksumResult,
+    cli_algo_name: Option<&str>,
+    cli_algo_length: Option<usize>,
+    properly_formatted: &mut bool,
+    opts: ChecksumOptions,
+) -> Result<bool, LineCheckError> {
```

Comment:
Why does the function return a `bool` in the success case?

Reply:
It's not a hard requirement. I just wanted to represent the Result as an answer to "could the digest comparison happen ?". So either everything is right and the comparison is done, thus returning `Ok()` with the result of the comparison, or the process failed beforehand and the `Err` is describing why.

Reply:
Hm, is the "could the digest comparison happen?" question even relevant in this context? Maybe I'm missing something.

Reply:
I decided to go with this approach, but it's not a hard requirement. I will change the return type of the function to `Result<(), LineCheckError>` and add an error variant for failed digest comparison.